### PR TITLE
Add subnet tag controller for subnet tags

### LINKED ISSUE #692: Implement tags managed resource
### What problem are you facing?
There is currently no managing of ec2 tags as crossplane objects, which would allow to set tags on aws resources. 

This would come in handy with the [recent change](https://docs.aws.amazon.com/eks/latest/userguide/network_reqs.html#vpc-subnet-tagging) in EKS, where AWS doesn't tag the respective subnets automatically.

### How could Crossplane help solve your problem?
Implement [CreateTags](https://pkg.go.dev/github.com/aws/aws-sdk-go@v1.38.52/service/ec2#CreateTags), [DeleteTags](https://pkg.go.dev/github.com/aws/aws-sdk-go@v1.38.52/service/ec2#DeleteTags) via ACK code generation

### PR DESCRIPTION
**Pull request by MisterMX**
### Description of your changes

Fixes #692 

This PR adds a new `ResourceTag` resource that allows tagging an AWS EC2 resource without having to rely on the respective Crossplane resource directly.

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

Manually and unit testing

### CODE REVIEW

**Comment by MisterMX** [comment]
Updated this PR to a generic `ResourceTag` instead of the specific `SubnetTag` resource to better match the AWS API.

**Review by janwillies — CHANGES_REQUESTED** [review]
Thanks @MisterMX! I wonder if this API should be simply called `Tags`, since `ResourceTags` can be confused with the [Resource Groups Tagging API](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/resourcegroupstaggingapi)?

File: examples/ec2/subnettag.yaml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+apiVersion: ec2.aws.crossplane.io/v1alpha1
+kind: SubnetTag
```
[the remark this answers is not in the document]
Comment (MisterMX) [reply examples/ec2/subnettag.yaml]: No its a leftover from refactoring. Thanks for mentioning it. I corrected it.

File: apis/ec2/manualv1alpha1/referencers.go [hunk 2]
```diff
@@ -16,32 +16,22 @@ limitations under the License.
...
-		Extract:      reference.ExternalName(),
-	})
-	if err != nil {
-		return errors.Wrap(err, "spec.forProvider.vpcId")
-	}
-	mg.Spec.ForProvider.VPCID = reference.ToPtrValue(rsp.ResolvedValue)
-	mg.Spec.ForProvider.VPCIDRef = rsp.ResolvedReference
-	return nil
-}
+// func (mg *VPCCIDRBlock) ResolveReferences(ctx context.Context, c client.Reader) error {
```
[the remark this answers is not in the document]
Comment (MisterMX) [reply apis/ec2/manualv1alpha1/referencers.go]: No that a mistake by me. This part shouldn't be changed at all. I removed the changes.

File: apis/ec2/manualv1alpha1/resource_tag.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
+/*
+Copyright 2019 The Crossplane Authors.
```
[the remark this answers is not in the document]
Comment (MisterMX) [reply apis/ec2/manualv1alpha1/resource_tag.go]: Fixed

**Comment by MisterMX** [comment]
@janwillies I understand what you mean. The name is quite generic. Though the AWS API also just calls it `CreateTags` and `DeleteTags`. Do you have a suggestion for a less ambiguous name?

**Comment by janwillies** [comment]
Looking at the ec2 api I believe it should be `tags`, but I'll defer to @chlunde who has much more knowledge in this area.

File: apis/ec2/manualv1alpha1/register.go [hunk 4]
```diff
@@ -48,6 +48,15 @@ var (
 	VPCCIDRBlockGroupVersionKind = SchemeGroupVersion.WithKind(VPCCIDRBlockKind)
 )
 
+// SubnatTag type metadata.
```
Comment (haarchri) [inline apis/ec2/manualv1alpha1/register.go:51]: `// SubnetTag type metadata.`

**Comment by chlunde** [comment]
Assigning @negz / @muvaf because I think this needs a design review. A generic EC2 Tags resource does have some nice use cases, but it will cause issues if the resource is managed by another crossplane controller - and possibly also if it is managed by terraform or another tool, so I think we need to think this through.

**Comment by chlunde** [comment]
_just thinking out loud here:_

This resource type should strive to co-exist with other features (at least that might be a mode). So if another resource is tagged `{t0: foo, t1: foo}` and the user requests `{t2: faz}`, I think most people would want the resource to be tagged `{t0: foo, t1: foo, t2: faz}`. If you delete the resource, only the `t2` tag should be removed.

One way to achieve this would be to have a status field `ownedTags: [t2]`, meaning that we would only reconcile the tags we "own".

This might have to be put behind a flag under `spec`, because there might be other useful modes:
* take full ownership of tags using this controller.
* initially take full ownership, then move to "merge" mode
* merge mode from start
* define a tag prefix and take full ownership under that prefix

Possibly there should be a way to delete a tag (written by other tools) even in "merge" mode, it might be possible to use the nil/undefined tag value for this.

And this might work well with [terraform w/ ignore_changes](https://www.terraform.io/docs/language/meta-arguments/lifecycle.html#ignore_changes), but it does not work with crossplane. So maybe this `tag mode` think should exist for *all* (AWS) resources?

File: apis/ec2/manualv1alpha1/resource_tag.go [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+// ResourceTagParameters define the desired state of an AWS Resource ResourceTag.
+type ResourceTagParameters struct {
+	// Region is the region you'd like your Tag to be created in.
+	// +optional
+	Region *string `json:"region,omitempty"`
+
+	// IDs of the resources that should be tagged
+	// +optionial
```
Comment (negz) [inline apis/ec2/manualv1alpha1/resource_tag.go:43]:
```suggestion
	// +optional
```

File: apis/ec2/manualv1alpha1/resource_tag.go [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+// A ResourceTagSpec defines the desired state of a ResourceTag.
+type ResourceTagSpec struct {
+	xpv1.ResourceSpec `json:",inline"`
+	ForProvider       ResourceTagParameters `json:"forProvider"`
+}
+
+// A ResourceTagStatus represents the observed state of a ResourceTag.
+type ResourceTagStatus struct {
+	xpv1.ResourceStatus `json:",inline"`
+	// AtProvider          ResourceTagObservation `json:"atProvider,omitempty"`
```
Comment (negz) [inline apis/ec2/manualv1alpha1/resource_tag.go:61]:
```suggestion
```

**Comment by negz** [comment]
Thanks for the tag @chlunde. I agree this pattern needs discussion. It's a novel concept that I believe goes against how we've historically handled other resources in this provider - i.e. I believe we often model tags as a property of other managed resources (which is arguably a less faithful representation of the AWS API than the approach in this PR).

One shortcoming of this approach is that with our current reference resolution pattern I imagine it would be impossible to ever have an array of `resourceIdRef` or a `resourceIdsSelector`, because the referenced resources would be of arbitrary type (and iirc not all AWS resources can be tagged).

**Comment by MisterMX** [comment]
@chlunde  @negz many controllers follow the `add, remove` pattern and will delete any tags from the resource they don't know, except for `Subnet` which only every adds new tags to a resource - it also was the initial purpose of this PR before I rewrote it to be more generic (hence the title).

I really like the idea of a `tagMode` field which would make this resource much more flexible.

**Comment by MisterMX** [comment]
Since Subnet Tags are not needed anymore for load balancers this PR is obsolete and can be closed.

**Comment by haarchri** [comment]
we have a use-case for this PR
we need to Tag for Karpenter the EKS generated SecurityGroup 
https://karpenter.sh/v0.13.1/getting-started/migrating-from-cas/#add-tags-to-subnets-and-security-groups

```
NODEGROUP=$(aws eks list-nodegroups --cluster-name ${CLUSTER_NAME} \
    --query 'nodegroups[0]' --output text)

LAUNCH_TEMPLATE=$(aws eks describe-nodegroup --cluster-name ${CLUSTER_NAME} \
    --nodegroup-name ${NODEGROUP} --query 'nodegroup.launchTemplate.{id:id,version:version}' \
    --output text | tr -s "\t" ",")

SECURITY_GROUPS=$(aws ec2 describe-launch-template-versions \
    --launch-template-id ${LAUNCH_TEMPLATE%,*} --versions ${LAUNCH_TEMPLATE#*,} \
    --query 'LaunchTemplateVersions[0].LaunchTemplateData.SecurityGroupIds' \
    --output text)

aws ec2 create-tags \
    --tags "Key=karpenter.sh/discovery,Value=${CLUSTER_NAME}" \
    --resources ${SECURITY_GROUPS}
```

we can get the sg from `kubectl describe cluster.eks.aws.crossplane.io/crossplane-dev-dr9p7-98594` status.atProvider.resourcesVpcConfig.clusterSecurityGroupId: `sg-08c821268fffc30aa` but we have no chance without this PR to tag the SG 

in jet-provider we have the same resource https://doc.crds.dev/github.com/crossplane-contrib/provider-jet-aws/ec2.aws.jet.crossplane.io/Tag/v1alpha1@v0.4.0-preview#spec-forProvider 

for me +1 to go forward with Tag Resource

**Comment by haarchri** [comment]
we can add a few comments like terraform

`Manages an individual EC2 resource tag. This resource should only be used in cases where EC2 resources are created outside Crossplane (e.g., AMIs), being shared via Resource Access Manager (RAM), or implicitly created by other means (e.g., Transit Gateway VPN Attachments, EKS ClusterSecurityGroup).`

`This tagging resource should not be combined with the crossplane resource for managing the parent resource. For example, using vpc and tagresource to manage tags of the same VPC will cause a perpetual difference where the vpc resource will try to remove the tag being added by the tagresource.`

**Comment by zetamorph** [comment]
Hi 👋 
is there any chance we can re-open this? 
We have the same use case as the one @haarchri described [above](https://github.com/crossplane-contrib/provider-aws/pull/822#issuecomment-1170863716): we want to tag the security groups that are created automatically when EKS clusters are created.

Currently we're using provider-jet-aws with it's `Tag` resource to handle this.

**Comment by zetamorph** [comment]
Another possible use case is as follows:

EKS Managed Nodegroups each have an underlying EC2 Autoscaling group. We need to control the tags of these ASGs.
By using a Launchtemplate and configuring `tagSpecifications`, we can apply tags to the EC2 instances managed by the ASG. However, these tags are not applied to the ASG, just the instances.

This is a shortcoming of the AWS API, see also [this issue](https://github.com/aws/containers-roadmap/issues/608), but since it doesn't look like the issue will be resolved soon, the ability to tag arbitrary resources using `provider-aws` would be really helpful.

**Comment by MisterMX** [comment]
> is there any chance we can re-open this?

@zetamorph unfortunately, I don't have the time to work on this issue. But feel free reuse my changes if you have a use case.

**Pull request closed without merging**